cmd/ttn-lw-cli/internal/util/flags.go

```diff
@@ -548,7 +569,7 @@ func setField(rv reflect.Value, path []string, v reflect.Value) error {
 							break
 						}
 						var enum ttnpb.DataRateIndex
-						if err := enum.UnmarshalText([]byte(v.String())); err != nil {
+						if err := enum.UnmarshalText([]byte(v.String())); err != nil && enum != 0 {
```

Comment:
Long comment ahead. Relevant part of the PR description:

>  For custom enum values (e.g. RxDelayValue) passing a random string would count as a zero or nil value (e.g. --mac-settings.rx1-delay=random would work). This has also been fixed.

One of the relevant fields for that is `mac-settings.ping-slot-data-rate-index`. The accepted values for that field is:

```
 --mac-settings.ping-slot-data-rate-index string                   DATA_RATE_1|DATA_RATE_3|DATA_RATE_4|DATA_RATE_5|DATA_RATE_6|DATA_RATE_13|DATA_RATE_14|DATA_RATE_15|DATA_RATE_12|DATA_RATE_0|DATA_RATE_2|DATA_RATE_7|DATA_RATE_8|DATA_RATE_9|DATA_RATE_10|DATA_RATE_11
```

So, in order to set to 0, one has to set `--mac-settings.ping-slot-data-rate-index=DATA_RATE_0`

Now, **Without** the highlighted change you get the following behaviour:

```
# DATA_RATE_0 works ok, sets value to 0
$ ttn-lw-cli dev update app1 dev1 --mac-settings.ping-slot-data-rate-index=DATA_RATE_0 

# 0 prints an error, as expected according to the help text
$ ttn-lw-cli dev update app1 dev1 --mac-settings.ping-slot-data-rate-index=0
error:cmd/ttn-lw-cli/internal/util:flag_value (invalid flag value)
    mac-settings.ping-slot-data-rate-index=invalid value "0" for DataRateIndexValue
exit status 255

# any random string, instead of printing an error, will set the value to 0 instead
$ ttn-lw-cli dev update app1 dev1 --mac-settings.ping-slot-data-rate-index=aasdfasdfasfasdf
```

This is because `UnmarshalText` will return a valid enum value equal to 0 instead of throwing an error. https://github.com/TheThingsNetwork/lorawan-stack/blob/85767bc21f60409b24572fd523cba21a820eada4/cmd/ttn-lw-cli/internal/util/flags.go#L551

**With** the change, what we get is:

```
# DATA_RATE_0 works ok, sets value to 0
$ ttn-lw-cli dev update app1 dev1 --mac-settings.ping-slot-data-rate-index=DATA_RATE_0 

# 0 prints an error, as expected according to the help text
$ ttn-lw-cli dev update app1 dev1 --mac-settings.ping-slot-data-rate-index=0
error:cmd/ttn-lw-cli/internal/util:flag_value (invalid flag value)
    mac-settings.ping-slot-data-rate-index=invalid value "0" for DataRateIndexValue
exit status 255

# any random string also prints an error
$ ttn-lw-cli dev update app1 dev1 --mac-settings.ping-slot-data-rate-index=aasdfasdfasfasdf
```

Same things apply for the rest of the enum values, resolving the other conversations.

What is interesting, we also have:

```
      --mac-state.desired-parameters.rx2-data-rate-index string         0|1|10|11|12|13|14|15|2|3|4|5|6|7|8|9
```

In general, for the desired parameters, the numerical values are expected, instead of the `DATA_RATE_x` strings. This looks to be a difference because of `DataRateIndexValue` vs `DataRateIndex` types. Yet, the cli accepts both:

```
$ ttn-lw-cli dev update app1 dev1 --mac-state.desired-parameters.rx2-data-rate-index=1
$ ttn-lw-cli dev update app1 dev1 --mac-state.desired-parameters.rx2-data-rate-index=DATA_RATE_1
```

Considering the above, this is why I went with the proposed change. I am open to suggestions.

Reply:
#2258 also seems to relevant.

Reply:
1. `UnmarshalText` should return an error on invalid input, otherwise it's a bug and needs to be fixed. Giving a (very) quick look at `pkg/ttnpb/lorawan.go` I did not see anything wrong in this regard, but maybe I'm wrong.
2. We do need to support `0` as valid enum values for some enums. The only one which should not support `0` that I can think of now is `Rx1Delay`, but that, again, belongs in validation not actual unmarshaling logic.
3. Regarding numerical and string syntax, we *should* support both.

Reply:
1. Okay, so no bugs in `UnmarshalText`. Just noticed that the error actually lies in the `err != nil` condition, does it not? :dancer:
2. Agreed, fixing the condition, cause this was the actual issue, not allowing zero values.
3. Both are supported, it's just that the values are missing in the help text.